### PR TITLE
Containers: Implement get_image_url()

### PR DESCRIPTION
**Pull request by pdostal**
Implement `get_image_url()` instead of `get_container_image_to_test()`, `get_container_url_from_var()` and `get_suse_container_urls()`. We have a list of SUSE and openSUSE released and unreleased container images but we always need only one image url to work with.

- Related ticket: [poo#108058](https://progress.opensuse.org/issues/108058)
- Verification run: In further comments

### CODE REVIEW

File: lib/containers/urls.pm
Comment (jlausuch) [inline lib/containers/urls.pm:305]: I would prefer to have a name that implies an `image` and not a `container`.
e.g. `get_image_url` or `get_container_image_url`

**Comment by jlausuch** [comment]
Looks like it goes into the right direction. I left some suggestions. 
I guess you still have in mind to modify all the tests using `get_suse_container_urls` by this new method.

File: tests/containers/image.pm
Comment (b10n1k) [inline tests/containers/image.pm:30]: is this module something which is scheduled in both `released`, `unreleased`? With current approach you will have always `released=1`. i think this breaks for instance https://openqa.suse.de/tests/8312796#step/image_docker/46

File: lib/containers/urls.pm
Comment (b10n1k) [inline lib/containers/urls.pm:321]:
```suggestion
   my $repo = ($args{released}) ? 'released' : totest;
   return $images_list{$args{distri}}{$args{version}}{$repo}->($args{arch});  if supports_image_arch($args{distri}, $args{version}, $args{arch});
```

**Review by b10n1k — CHANGES_REQUESTED** [review]
i think `$args{released}` will cause some issues in the modules which is used in both _released_ and _totest_. One option is to make use of the already existing variables `CONTAINER_IMAGE_TYPE` or `CONTAINERS_UNTESTED_IMAGES` i guess

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm:321]: Sorry, I will not do that. It just decreases readability as the lines would be too long.

File: tests/containers/image.pm
Comment (pdostal) [reply tests/containers/image.pm:30]: Good catch, thank you :smile:

File: tests/containers/container_diff.pm
Comment (pdostal) [reply tests/containers/container_diff.pm:24]: Good catch, thank you :smile:

File: lib/containers/urls.pm
Comment (b10n1k) [reply lib/containers/urls.pm:320]: Sorry i reopen it for further discussion. its a small thing but how does it decreases readability? the main issue here is that you have pretty much the same return twice and i cant see any good reason.

Comment (pdostal) [reply lib/containers/urls.pm:320]: Readability is the reason.

Comment (pdostal) [reply lib/containers/urls.pm:320]: I am not sure I understand. Do you have another suggestion?

Comment (b10n1k) [reply lib/containers/urls.pm:320]: it depends what do you consider /readability/

Comment (jlausuch) [reply lib/containers/urls.pm:320]: Agree, that 1 liner is nice but difficult to understand at a first glance...  maybe... 
```
my $type = $args{released} ? 'released' : 'totest';
if supports_image_arch($args{distri}, $args{version}, $args{arch}) {
    return $images_list{$args{distri}}{$args{version}}{$type}->($args{arch})
}
```

**Comment by jlausuch** [comment]
> i think `$args{released}` will cause some issues in the modules which is used in both _released_ and _totest_. One option is to make use of the already existing variables `CONTAINER_IMAGE_TYPE` or `CONTAINERS_UNTESTED_IMAGES` i guess

Exactly, `CONTAINER_IMAGE_TYPE` doesn't exist. `CONTAINERS_UNTESTED_IMAGES=1` means that we should use the `totest` flag.

File: lib/containers/urls.pm
Comment (jlausuch) [inline lib/containers/urls.pm:309]:
```suggestion
    $args{released} //= ! get_var('CONTAINERS_UNTESTED_IMAGES');
```

File: lib/containers/urls.pm
Comment (jlausuch) [inline lib/containers/urls.pm:322]: It's time to change this error message. To me, when I hit this error in openQA, it was difficult to understand why and I needed to dig into the code. So, for someone who is not used to this code, it's a bit tricky to understand. 

Also, ternary operators are awesome, but when the line is too long, it makes the code a bit unreadable.
So, either you use if-else clause, or store the message in a variable.
Suggestion:
```
my $error_msg = "Cannot find container image for $args{distri} $args{version} $args{arch} or missing CONTAINER_IMAGE_TO_TEST variable";

($args{die}) ? die($error_msg ) : return undef;
```

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm:314]: The `trim()` is there to trim spaces at the beginning or end of the variable.
E.g. if you specify `CONTAINER_IMAGE_TO_TEST="registry.suse.de/sle:15.3 "` it will *trim* the last space.

This was suggested by @ilausuch - Ivan, can you help here? :smile:

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm]: I like this too, thank you @jlausuch !

File: lib/containers/urls.pm
Comment (jlausuch) [reply lib/containers/urls.pm:313]: Spaces don't harm. But the trim will just simply fail if the variable is empty, which is wrong approach.
If we want to do checks, I would do it in the right way, using `URI` package from perl, e.g. https://stackoverflow.com/questions/2599047/how-can-i-validate-a-website-url-in-perl
But that's optional, that can be improved later, I would simple go for the code I suggested.

**Review by b10n1k — APPROVED** [review]

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm]: I see your point. But I wonder how is then possible that my example worker in all the verification runs both with and without this variable.

Please re-check.

File: lib/containers/urls.pm
Comment (grisu48) [inline lib/containers/urls.pm:301]:
```suggestion
# Return URI of the container image to be tested. This can be either configured via the CONTAINER_IMAGE_TO_TEST variable or it will try to determinte the same image as the running OS. Dies if the image could not be determined
```

WDYT?

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm]: Yes. I removed the `trim` - is it resolved?

File: tests/containers/container_diff.pm
Comment (grisu48) [inline tests/containers/container_diff.pm:29]: `docker pull` smells like a good candidate for `script_retry` ....

```suggestion
    script_retry("docker pull $unreleased_image", timeout => 360, retries => 3, delay => 120);
```

Same for the next line

**Comment by pdostal** [comment]
After I get some more approvals I'd like to do more verification runs.

**Review by grisu48 — COMMENTED** [review]
Good job Pavel! I added some suggestions, but feel free to ignore them if not applicable.

File: lib/containers/urls.pm
Comment (jlausuch) [reply lib/containers/urls.pm:301]: There are 3 cases:

- Provide `CONTAINER_IMAGE_TO_TEST` variable with URI
- Give it as input parameter `$args{version}` which is what we do to test our big matrix.
- Determine the same image as the running OS.

File: tests/containers/container_diff.pm
Comment (pdostal) [reply tests/containers/container_diff.pm:29]: Yes, let's leave this for separate pull request please.

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm:301]:
```
# Return URI of the container image to be tested. This will be:
# - Configured by the CONTAINER_IMAGE_TO_TEST variable.
# - Configured by the distri, version and arch parameter.
# - Determined by the running OS.
# - Terminated otherwise.
```

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm]: I implemented Ivan's solution, thank you!
```
$url = get_var('CONTAINER_IMAGE_TO_TEST');
return trim($url) if ($url)
```

File: lib/containers/urls.pm
Comment (pdostal) [reply lib/containers/urls.pm]: Please re-check.

File: lib/containers/urls.pm
Comment (jlausuch) [reply lib/containers/urls.pm]: Sorry for insisting, but I still don't understand why we need trim here :) 
Just to remove any spaces in the variable is not a good reason, we would need to do this for every variable, which doesn't make sense.

**Review by grisu48 — APPROVED** [review]
LGTM

**Review by jlausuch — COMMENTED** [review]
Looks very good 👍 
Can you please run more VRs?

File: tests/containers/rootless_podman.pm
Comment (pdostal) [reply tests/containers/rootless_podman.pm:62]: :smile: thank you

**Comment by pdostal** [comment]
Verification runs: [15-SP3 on AKS](http://pdostal-server.suse.cz/tests/13876), [15-SP3-docker](http://pdostal-server.suse.cz/tests/13879), [15-SP3-podman](http://pdostal-server.suse.cz/tests/13880), [Tumbleweed-podman](http://pdostal-server.suse.cz/tests/13875)

**Review by ilausuch — APPROVED** [review]

**Comment by ilausuch** [comment]
Mice work!

**Pull request merged**
